File: bot.js
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { URL, URLSearchParams } = require('url');
const fetch = require('node-fetch');

const CONFIG_PATH = path.join(__dirname, 'config.json');
const TOKENS_PATH = path.join(__dirname, 'tokens.json');

const TOKEN_URL = 'https://api.twitter.com/2/oauth2/token';
const TWEET_URL = 'https://api.twitter.com/2/tweets';

// Gemini (Generative Language API)
function geminiGenerateUrl(model, apiKey) {
  return `https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(model)}:generateContent?key=${apiKey}`;
}

const TEST_SENTENCES = [
  'Testing automated post 1: Hello from my hourly bot!',
  'Testing automated post 2: This is a scheduled tweet.',
  'Testing automated post 3: Verifying OAuth refresh works.',
  'Testing automated post 4: Rotating through messages.',
  'Testing automated post 5: Everything looks good so far!'
];

function loadConfig() {
  let fileCfg = {};
  try {
    if (fs.existsSync(CONFIG_PATH)) {
      fileCfg = JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf8'));
    }
  } catch (_) {
    fileCfg = {};
  }
  const merged = {
    ...fileCfg,
    // OAuth 1.0a credentials
    api_key: process.env.API_KEY || fileCfg.api_key,
    api_secret: process.env.API_SECRET || fileCfg.api_secret,
    access_token: process.env.ACCESS_TOKEN || fileCfg.access_token,
    access_token_secret: process.env.ACCESS_TOKEN_SECRET || fileCfg.access_token_secret,

    // Gemini
    gemini_api_key: process.env.GEMINI_API_KEY || fileCfg.gemini_api_key,
    gemini_model: process.env.GEMINI_MODEL || fileCfg.gemini_model || 'gemini-1.5-flash',

    // Interval
    post_every_hours: Number(process.env.POST_EVERY_HOURS || fileCfg.post_every_hours || 1),

    // OAuth2 (if used locally)
    client_id: process.env.CLIENT_ID || fileCfg.client_id,
  };
  return merged;
}

function loadTokens() {
  return JSON.parse(fs.readFileSync(TOKENS_PATH, 'utf8'));
}

function saveTokens(tokens) {
  fs.writeFileSync(TOKENS_PATH, JSON.stringify(tokens, null, 2), 'utf8');
}

// =========================
// Gemini text generation
// =========================
async function generateTweet(cfg) {
  if (!cfg.gemini_api_key) {
    // Not configured, fallback
    return TEST_SENTENCES[Math.floor(Math.random() * TEST_SENTENCES.length)];
  }

  const model = cfg.gemini_model || 'gemini-1.5-flash';
  const url = geminiGenerateUrl(model, cfg.gemini_api_key);

  // Rotating on-brand prompts for Wenlambo.lol
  const PROMPTS = [
    'Wenlambo.lol lets anyone launch a new token or register an existing one — free to create, just pay gas. Tokens trade instantly on our marketplace. Creators earn 5% at launch + 0.1% per swap auto-routed. Add 1-3 crypto emojis + 2-4 crypto hashtags.',
    'Launch or register your token on Wenlambo.lol in minutes. No platform fees (gas only), instant trading, and creator rewards: 5% of total supply at launch and 0.1% on every swap to the creator. Use 1-3 emojis and 2-4 crypto hashtags.',
    'Ship your token on Wenlambo.lol: free creation (gas only), instant marketplace trading, and built-in incentives — 5% supply at launch + 0.1% of all swaps to creators. Include 1-3 emojis and 2-4 crypto hashtags.',
    'Register an existing token or launch a new one on Wenlambo.lol. Creation/registration are free (gas only). Trade immediately. Creators get 5% at launch and 0.1% per swap automatically. Add 1-3 emojis + 2-4 crypto hashtags.'
  ];
  const prompt = PROMPTS[Math.floor(Math.random() * PROMPTS.length)] + ' Output a single tweet under 240 characters, include 2-4 relevant crypto hashtags and 1-3 emojis. No URLs. Output only the tweet text.';

  const body = {
    contents: [
      { parts: [{ text: prompt }] }
    ],
    generationConfig: {
      temperature: 0.9,
      maxOutputTokens: 120
    }
  };

  const resp = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  });

  if (!resp.ok) {
    const t = await resp.text();
    throw new Error(`Gemini error: ${resp.status} ${t}`);
  }

  const data = await resp.json();
  let text = null;
  if (data && Array.isArray(data.candidates) && data.candidates[0] && data.candidates[0].content && Array.isArray(data.candidates[0].content.parts)) {
    // Concatenate all parts text if multiple
    text = data.candidates[0].content.parts.map(p => p.text || '').join(' ').trim();
  }
  if (!text) throw new Error('Gemini returned no content');

  // Sanitize to a single-line tweet and cap to 280 chars (Twitter limit)
  text = String(text).trim().replace(/\s+/g, ' ');
  text = text.replace(/^"|"$/g, '').replace(/^`+|`+$/g, '');
  if (text.length > 280) text = text.slice(0, 279);

  return text;
}

// =========================
// OAuth 1.0a (User context)
// =========================
function hasOAuth1(cfg) {
  return Boolean(cfg.api_key && cfg.api_secret && cfg.access_token && cfg.access_token_secret);
}

function rfc3986Encode(str) {
  return encodeURIComponent(str)
    .replace(/[!'()*]/g, c => '%' + c.charCodeAt(0).toString(16).toUpperCase());
}

function buildBaseString(method, baseURL, params) {
  // params: object of key->array(values) already normalized to arrays
  const pairs = [];
  const keys = Object.keys(params).sort();
  for (const k of keys) {
    const vs = Array.isArray(params[k]) ? params[k] : [params[k]];
    vs.sort();
    for (const v of vs) {
      pairs.push(`${rfc3986Encode(k)}=${rfc3986Encode(String(v))}`);
    }
  }
  const paramString = pairs.join('&');
  const base = [method.toUpperCase(), rfc3986Encode(baseURL), rfc3986Encode(paramString)].join('&');
  return base;
}

function oauth1HeaderForRequest({ method, url, cfg }) {
  const parsed = new URL(url);
  const baseURL = `${parsed.origin}${parsed.pathname}`;

  const oauthParams = {
    oauth_consumer_key: cfg.api_key,
    oauth_nonce: crypto.randomBytes(16).toString('hex'),
    oauth_signature_method: 'HMAC-SHA1',
    oauth_timestamp: Math.floor(Date.now() / 1000).toString(),
    oauth_token: cfg.access_token,
    oauth_version: '1.0',
  };

  // Collect all parameters (OAuth + query) for signature base string
  const sigParams = { ...oauthParams };
  for (const [k, v] of parsed.searchParams.entries()) {
    if (sigParams[k] === undefined) sigParams[k] = [];
    if (!Array.isArray(sigParams[k])) sigParams[k] = [sigParams[k]];
    sigParams[k].push(v);
  }

  const signingKey = `${rfc3986Encode(cfg.api_secret)}&${rfc3986Encode(cfg.access_token_secret)}`;
  const baseString = buildBaseString(method, baseURL, sigParams);
  const signature = crypto.createHmac('sha1', signingKey).update(baseString).digest('base64');

  const headerParams = { ...oauthParams, oauth_signature: signature };
  const auth = 'OAuth ' + Object.keys(headerParams)
    .sort()
    .map(k => `${rfc3986Encode(k)}="${rfc3986Encode(headerParams[k])}"`)
    .join(', ');
  return auth;
}

async function postTweetOAuth1(cfg, text) {
  const method = 'POST';
  const url = TWEET_URL; // no query params
  const headers = {
    'Authorization': oauth1HeaderForRequest({ method, url, cfg }),
    'Content-Type': 'application/json',
  };
  const resp = await fetch(url, {
    method,
    headers,
    body: JSON.stringify({ text })
  });
  if (!(resp.status === 200 || resp.status === 201)) {
    const textResp = await resp.text();
    throw new Error(`Failed to post tweet (OAuth1): ${resp.status} ${textResp}`);
  }
  return await resp.json();
}

// =========================
// OAuth 2.0 (PKCE user flow)
// =========================
async function ensureToken(cfg, tokens) {
  const now = Math.floor(Date.now() / 1000);
  if (tokens.access_token && Number(tokens.expires_at || 0) > now) {
    return tokens;
  }
  console.log('Refreshing access token...');
  const params = new URLSearchParams({
    grant_type: 'refresh_token',
    refresh_token: tokens.refresh_token,
    client_id: cfg.client_id,
  });
  const resp = await fetch(TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: params.toString(),
  });
  if (!resp.ok) {
    const text = await resp.text();
    throw new Error(`Failed to refresh token: ${resp.status} ${text}`);
  }
  const newTok = await resp.json();
  const expiresIn = Number(newTok.expires_in || 0);
  newTok.expires_at = Math.floor(Date.now() / 1000) + expiresIn - 60;
  if (!newTok.refresh_token && tokens.refresh_token) {
    newTok.refresh_token = tokens.refresh_token;
  }
  saveTokens(newTok);
  console.log('Token refreshed.');
  return newTok;
}

async function postTweetOAuth2(accessToken, text) {
  const resp = await fetch(TWEET_URL, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${accessToken}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ text })
  });
  if (!(resp.status === 200 || resp.status === 201)) {
    const textResp = await resp.text();
    throw new Error(`Failed to post tweet: ${resp.status} ${textResp}`);
  }
  return await resp.json();
}

async function main() {
  const cfg = loadConfig();
  const intervalHours = Number(cfg.post_every_hours || 1);
  const intervalMs = Math.max(60_000, intervalHours * 3600_000);

  const useOAuth1 = hasOAuth1(cfg);

  if (!useOAuth1) {
    // OAuth2 path requires tokens.json
    if (!fs.existsSync(TOKENS_PATH)) {
      console.error('No OAuth 1.0a credentials found in config and tokens.json not found.');
      console.error('Either:');
      console.error('  1) Provide api_key, api_secret, access_token, access_token_secret in config.json (OAuth 1.0a), or');
      console.error('  2) Run "npm run auth" to authorize with OAuth 2.0 (PKCE).');
      process.exit(1);
    }
  }

  let tokens = null;
  if (!useOAuth1) {
    tokens = loadTokens();
  }

  async function cycle() {
    try {
      // Generate tweet via Gemini (with safe fallback)
      let text;
      try {
        text = await generateTweet(cfg);
      } catch (genErr) {
        console.error('Gemini generation failed, using fallback:', genErr.message || genErr);
        text = TEST_SENTENCES[Math.floor(Math.random() * TEST_SENTENCES.length)];
      }

      let resp;
      console.log('Posting tweet:', text);
      if (useOAuth1) {
        resp = await postTweetOAuth1(cfg, text);
      } else {
        tokens = await ensureToken(cfg, tokens);
        resp = await postTweetOAuth2(tokens.access_token, text);
      }
      console.log('Tweet posted:', resp);
    } catch (err) {
      console.error('Error during posting:', err);
    }
  }

  // Run immediately then on interval
  await cycle();
  setInterval(cycle, intervalMs);
}

main().catch((err) => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: bot.py
import os
import time
import random
import hashlib
import json
import re
from datetime import datetime, timezone
from typing import Optional

import requests
from requests_oauthlib import OAuth1

# ========================= Config =========================
API_KEY = os.getenv("API_KEY")
API_SECRET = os.getenv("API_SECRET")
ACCESS_TOKEN = os.getenv("ACCESS_TOKEN")
ACCESS_TOKEN_SECRET = os.getenv("ACCESS_TOKEN_SECRET")

# Gemini
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
GEMINI_MODEL = os.getenv("GEMINI_MODEL", "gemini-2.5-flash-lite")
GEMINI_URL = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent"

POST_EVERY_HOURS = int(os.getenv("POST_EVERY_HOURS", 3))  # tweet every 3 hours
LAUNCHPAD_NAME = os.getenv("LAUNCHPAD_NAME", "wenlambo")
LAUNCHPAD_WEBSITE = os.getenv("LAUNCHPAD_WEBSITE", "wenlambo.lol")

DATA_DIR = os.getenv("DATA_DIR", "data")
HISTORY_PATH = os.getenv("HISTORY_PATH", os.path.join(DATA_DIR, "posted_history.jsonl"))
WEBSITE_USAGE_PATH = os.getenv("WEBSITE_USAGE_PATH", os.path.join(DATA_DIR, "website_usage.json"))
PROMPT_INDEX_PATH = os.getenv("PROMPT_INDEX_PATH", os.path.join(DATA_DIR, "prompt_index.json"))

TWEET_URL = "https://api.twitter.com/2/tweets"

# ==========================================================

PROMPTS = [
    "short tweet that highlights what wenlambo launchpad offers, like free token creation and registration of tokens created outside the platform, 0.1% fees for creators, weekly airdrops, staking, and many more. Make it very convincing, fun, and full of energy with natural emojis.",
    "craft a viral tweet hyping how wenlambo is the launchpad for degens tired of high fees and complex tools. compare it with typical crypto pain points and show how easy it is to mint, trade, and stake instantly. keep it fun and punchy.",
    "make a meme-style crypto tweet comparing the chaos of rugpulls and gas wars to how simple it feels trading on a good launchpad. add crypto slang, humor, and energy. short, fun, natural.",
    "write a degen humor tweet about how everyone in crypto is waiting for the next moonshot, but only smart traders find the early gems. keep it fun, sarcastic, and engaging.",
    "ask the crypto community: which memecoin is the best right now? which one will actually moon next? make it hype and engaging, encouraging replies. fun degen-style with emojis.",
    "create a crypto culture tweet referencing trends like memecoins, staking, and airdrops. make it feel like a community vibe, fun and fast-paced.",
    "write a hype crypto tweet about a launchpad that gives free token creation, 0.1% swap royalties for creators, weekly airdrops, 5% supply to users, staking rewards & whale protection. Make it fun, human, with emojis and spacing.",
    "make a meme-style tweet comparing other launchpads that charge fees vs ours that pays users and creators. Add emojis, spacing, and degen slang. It should sound like a human shill, not a robot.",
    "create a short viral tweet that highlights how our launchpad is fair, rewarding, and safe from whales. Use hype energy, emojis, and spacing to sound like a community-driven crypto post.",
    "write a playful crypto tweet that flexes how users on our launchpad get airdrops every week + 5% supply at launch. Make it funny, human-like, with emojis and casual tone."
]

CRYPTO_HASHTAGS = [
    "#crypto", "#bnb", "#ethereum", "#bitcoin", "#memecoin",
    "#blockchain", "#web3", "#trading", "#altcoins", "#defi"
]

# ========================= Helpers =========================

def ensure_dir_for_file(path: str):
    d = os.path.dirname(path)
    if d and not os.path.exists(d):
        os.makedirs(d, exist_ok=True)

def normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", (text or "").strip()).lower()

def clean_spacing(text: str) -> str:
    text = re.sub(r"\s+([?.!,])", r"\1", text)
    text = re.sub(r"([?.!,])([^\s])", r"\1 \2", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text

def _trim_to_tweet(text: str, limit: int = 280) -> str:
    text = (text or "").strip()
    if len(text) <= limit:
        return text
    return text[: limit - 1].rstrip() + "…"

def append_history_record(text: str, tweet_ids: list):
    ensure_dir_for_file(HISTORY_PATH)
    has_website = LAUNCHPAD_WEBSITE.lower() in normalize_text(text)
    record = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "preview": normalize_text(text)[:200],
        "tweet_ids": tweet_ids,
        "has_website": has_website,
    }
    with open(HISTORY_PATH, "a", encoding="utf-8") as f:
        f.write(json.dumps(record, ensure_ascii=False) + "\n")

def _today_str() -> str:
    return datetime.now(timezone.utc).date().isoformat()

def _load_site_usage() -> dict:
    try:
        with open(WEBSITE_USAGE_PATH, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return {}

def should_include_site_today() -> bool:
    usage = _load_site_usage()
    return usage.get("date") != _today_str() or not usage.get("used", False)

def mark_site_used_today():
    ensure_dir_for_file(WEBSITE_USAGE_PATH)
    with open(WEBSITE_USAGE_PATH, "w", encoding="utf-8") as f:
        json.dump({"date": _today_str(), "used": True}, f)

# ================= Prompt Rotation =================

def load_prompt_index() -> int:
    try:
        with open(PROMPT_INDEX_PATH, "r", encoding="utf-8") as f:
            return int(json.load(f).get("index", 0))
    except Exception:
        return 0

def save_prompt_index(index: int):
    ensure_dir_for_file(PROMPT_INDEX_PATH)
    with open(PROMPT_INDEX_PATH, "w", encoding="utf-8") as f:
        json.dump({"index": index}, f)

def next_prompt() -> str:
    idx = load_prompt_index()
    seed = PROMPTS[idx % len(PROMPTS)]
    save_prompt_index((idx + 1) % len(PROMPTS))
    return seed

# ===================== Text Generation =====================

def build_viral_prompt(launchpad_name: str, website: str, seed: str, include_site: bool) -> str:
    rules = [
        "You are crafting a SINGLE viral-style crypto tweet. It must be punchy, humorous or mildly controversial, and persuasive.",
        f"Site: {website}.",
        "- Output ONE tweet only.",
        "- Max 260 characters.",
    ]
    if include_site:
        rules.append(f"- Include '{website}' exactly once.")
    else:
        rules.append(f"- Do NOT include '{website}'.")
    return "\n".join(rules) + f"\nTheme seed: {seed}."

def gemini_generate_text(prompt: str) -> str:
    headers = {"Content-Type": "application/json"}
    payload = {"contents": [{"parts": [{"text": prompt}]}]}
    r = requests.post(
        f"{GEMINI_URL}?key={GEMINI_API_KEY}",
        headers=headers,
        json=payload,
        timeout=60,
    )
    r.raise_for_status()
    data = r.json()
    return data["candidates"][0]["content"]["parts"][0]["text"].strip()

def ensure_brand_and_site(text: str, launchpad_name: str, website: str, include_site: bool) -> str:
    t = re.sub(r"\s+", " ", (text or "").strip())
    if include_site and website.lower() not in t.lower():
        t = _trim_to_tweet(t + f" — {website}", 280)
    return t

def add_crypto_hashtags(text: str, min_tags: int = 3) -> str:
    hashtags_in_text = [tag for tag in CRYPTO_HASHTAGS if tag.lower() in text.lower()]
    needed = max(0, min_tags - len(hashtags_in_text))
    if needed > 0:
        extra = random.sample(CRYPTO_HASHTAGS, needed)
        text = _trim_to_tweet(text + " " + " ".join(extra), 280)
    return text

def finalize_tweet(text: str, launchpad_name: str, website: str, include_site: bool) -> str:
    if (text.startswith("\"") and text.endswith("\"")) or (text.startswith("'") and text.endswith("'")):
        text = text[1:-1]
    text = ensure_brand_and_site(text, launchpad_name, website, include_site)
    text = clean_spacing(text)
    text = add_crypto_hashtags(text, min_tags=3)
    return _trim_to_tweet(text, 280)

def generate_viral_tweet(launchpad_name: str, website: str, include_site: bool) -> Optional[str]:
    seed = next_prompt()
    prompt = build_viral_prompt(launchpad_name, website, seed, include_site)
    try:
        text = gemini_generate_text(prompt)
    except Exception as e:
        print("Gemini text generation failed", e)
        return None
    if not text:
        return None
    return finalize_tweet(text, launchpad_name, website, include_site)

# ======================== Twitter API ========================

def post_tweet(text: str):
    auth = OAuth1(API_KEY, API_SECRET, ACCESS_TOKEN, ACCESS_TOKEN_SECRET)
    r = requests.post(TWEET_URL, auth=auth, json={"text": text}, timeout=60)
    if r.status_code not in (200, 201):
        raise RuntimeError(f"Failed to post tweet: {r.status_code} {r.text}")
    return r.json()

# =========================== Main ===========================

def main():
    if not all([API_KEY, API_SECRET, ACCESS_TOKEN, ACCESS_TOKEN_SECRET]):
        raise RuntimeError("Twitter API credentials not fully set.")
    if not GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY must be set.")

    interval_seconds = max(60, POST_EVERY_HOURS * 3600)

    while True:
        try:
            include_site = should_include_site_today()
            text = generate_viral_tweet(LAUNCHPAD_NAME, LAUNCHPAD_WEBSITE, include_site=include_site)
            if not text:
                print("⚠️ No tweet generated this cycle.")
                time.sleep(interval_seconds)
                continue
            print(f"🚀 Posting tweet (len={len(text)}, include_site={include_site}): {text}")
            try:
                resp = post_tweet(text)
                tweet_id = resp.get("data", {}).get("id")
                if include_site and LAUNCHPAD_WEBSITE.lower() in normalize_text(text):
                    mark_site_used_today()
                append_history_record(text, [tweet_id] if tweet_id else [])
                print("✅ Tweet posted. ID:", tweet_id)
            except Exception as e:
                print("❌ Failed to post:", e)
        except Exception as e:
            print("🔥 Error in loop:", e)
        print(f"😴 Sleeping for {interval_seconds} seconds...")
        time.sleep(interval_seconds)

if __name__ == "__main__":
    main()






